design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,890 @@
+  In practice, one will have crates with sole purpose to define message types.
+  So a consumer of messages, don't need to know any sources of such messages.
+  And vice-versa, the producers don't have to depend (at the code level) on any consumers.
+  Both consumers and producers only need a dependency on the message definitions.
+* The thin-edge API should give a large flexibility to connect actors:
+  * exchanging stream of messages as for measurements,
+  * sending requests which responses are awaited by the requester,
+  * sending asynchronous requests which responses will be processed concurrently with the other messages,
+  * broadcasting messages,
+  * gathering messages from various sources,
+  * sending messages to one specific instance of an actor (as a request to the apt plugin).
+* The final executable is build as an assemblage of extensions, defining actor instances and their connections.
+* This must not be the sole way to create MQTT-based thin-edge executables - aka *plugins*.
+  An agent developer must be free choosing his preferred programming language and design
+  to implement a feature that interacts with other thin-edge executables using JSON over MQTT.
+
+One should be able to build executables with IIoT specific features.
+* The `tedge_actor` crate defines only messages related to the runtime as `SpawnTask`, `Shutdown` or `Timeout`.
+* IIoT related messages are defined in specific extensions as `tedge_telemetry` or `tedge_software_management`.
+* An extension must be provided for MQTT as well as another one for HTTP.
+* An extension must be provided to encapsulate the thin-edge MQTT API,
+  with the definition of all the MQTT topics and message payloads.
+
+One should be able to build batteries-included executables
+that contains numerous extensions and alternative implementations for a large diversity of use-cases.
+
+* An extension can be included in the final executable without being enabled, but only registered and ready to be used.
+* Registered extensions can be enabled and configured at runtime.
+* Having extensions included but not instantiated should not impact other extensions nor consume runtime resources.
+* Such an executable with enabled and non-enabled extensions must provide a command line option, say `--list-extensions`,
+  to provide the whole list of available extensions and their purpose;
+  as well as command line options for detailed help on how to enable and configure those.
+
+Robustness is key.
+* One extension panicking should not impact other extensions or the runtime.
+* All errors must be handled in a non-crashing way.
+* Unrecoverable errors may cause the binary to shutdown eventually, but not unexpectedly.
+* The framework must handle SIGTERM and signal a shutdown to all active actors.
+* Shutdown is signalled to all extensions, giving them a possibility to handle such case gracefully.
```

Comment:
So would this support the following?
* Shutdown is signalled, all components are signaled and stop posting new messages to the queue, and the remaining messages in the queue are drained before the soft shutdown (though limiting how much time this takes to avoid a hard shutdown via `SIGKILL`).

Or are the actor messages fairly light weight, and any publishing of outgoing mqtt messages to broker are persisted by the broker itself?

Motivation here is to publish as any outgoing mqtt messages before the services or device is shutdown so that the cloud side has the latest information about the device (in case if the device has problems coming back online)

Reply:
What I have in mind is:
* The shutdown is triggered by an actor that listen to `SIGTERM`.
* On shutdown, the runtime sends a shutdown to all running actors.
* Each actor do a shutdown its own way, ignoring pending messages or not, stopping to send messages or not.
   * For instance, the MQTT actor might continue to publish all pending messages, while a measurement source might simply stop emitting.
* After some time, the runtime aborts any actor still running.

Reply:
Added some clarifications